pytorch_toolkit/ote/ote/modules/config_transformers/config_transformers_engine.py

```diff
@@ -0,0 +1,104 @@
+import datetime
+from tempfile import NamedTemporaryFile
+from copy import copy
+
+from ote.utils import load_config
+from ..builder import build_config_transformer
+
+def save_config(config, file_path):
+    with open(file_path, 'w') as output_stream:
+        yaml.dump(config, output_stream)
+
+class _ConfigTransformersHandler:
+    def __init__(self, template_path, config_transformers_names):
+        self.template_path = template_path
+        self.timestamp = self._generate_timestamp_suffix()
+
+        if config_transformers_names is None:
+            self.config_transformers = None
+            return
+
+        if not isinstance(config_transformers_names, list):
+            config_transformers_names = [config_transformers_names]
+        self.config_transformers = []
+        for cfg_transform_name in config_transformers_names:
+            self.config_transformers.append(build_config_transformer(cfg_transform_name))
+
+    def __call__(self, config_path):
+        if not self.config_transformers:
+            return config_path
+        logging.debug(f'_ConfigTransformersHandler: config_path={config_path}')
+        assert os.path.exists(config_path), f'The initial config path {config_path} is absent'
+        prev_config_path = config_path
+        for index, config_transformer in enumerate(self.config_transformers):
+            generated_config_path = self._generate_config_path(config_path, index)
+            logging.debug(f'_ConfigTransformersHandler: index={index}, generated_config_path={generated_config_path}')
+            assert not os.path.exists(generated_config_path), f'During generating configs path {generated_config_path} is present'
+            cfg_update_part = config_transformer(template_path=self.template_path,
+                                                 config_path=prev_config_path)
+
+            assert isinstance(cfg_update_part, dict), (
+                    f'Error in config transformer #{index} "{config_transformer}": it returns a value that is not a dict:'
+                    f' type(cfg_update_part)={type(cfg_update_part)}')
+            assert '_base_' not in cfg_update_part, (
+                    f'Error in config transformer #{index} "{config_transformer}": it returns a dict with key "_base_"')
+
+            generated_config_dir = os.path.dirname(generated_config_path) #just to be on the safe side, indeed they should be in the same folder
+            cfg_update_part['_base_'] = os.path.relpath(prev_config_path, generated_config_dir)
+            save_config(cfg_update_part, generated_config_path)
+
+            assert os.path.exists(generated_config_path), f'Cannot write config file "{generated_config_path}"'
+
+            prev_config_path = generated_config_path
+
+        return generated_config_path
+
+
+
+    @staticmethod
+    def _generate_timestamp_suffix():
+        return datetime.datetime.now().strftime('%y%m%d%H%M%S')
+
+    def _generate_config_path(self, config_path, index):
```

Comment:
there is no need to create this config near the original one, you can use `tempfile`, can't you? And it is more thread safe.
Final configuration file is saved in workdir anyway.

Reply:
Ilya, I agree that timestamps is a bad idea from the point of thread-safety, etc      
But I prefer to keep all the temporary files in the same folder (even with temporary names) than store them in `/tmp/`.    
It will make debugging in the future much more clean.

Reply:
`tempfile` provides easy to use functionality to configure directory for temporary files.